### ISSUE TITLE
Integral tie-back in AtlasPlugin.cpp seems wrong - causes control glitches

### ISSUE THREAD

**Issue by osrf-migration** [osrf/drcsim#182]
**Original report ([archived issue](https://osrf-migration.github.io/drcsim-gh-pages/#!/osrf/drcsim/issues/182)) by John Hsu (Bitbucket: [hsu](https://bitbucket.org/%7B0a186eae-abf0-4514-a951-23db5eccc286%7D/), GitHub: [hsu](https://github.com/hsu)).**

----------------------------------------
Per [thread](http://answers.gazebosim.org/question/2309/integral-tie-back-in-atlasplugincpp-seems-wrong/),

The code at roughly [line 1066 in AtlasPlugin.cpp](https://github.com/osrf/drcsim/blob/b019587d439dac062ee99ed5a77189513f48ae44/ros/atlas_msgs/AtlasPlugin.cpp#L1066) appears to be wrong. Specifically, it generally will invert the `kiq_i` term on a step change and saturate in the opposite of the commanded direction.

If a negative step is applied that results in a clamped force then the code can result in a `kiq_i` that is clamped at the positive limit. This results in a joint excursion in the opposite of the desired direction.

This is not a desired behavior. I would recommend simply removing the problematic code.

~~~
    // lock integral term to provide continuous control as system moves
    // out of staturation
    this->errorTerms[i].k_i_q_i = math::clamp(
            this->errorTerms[i].k_i_q_i + (forceClamped - forceUnclamped),
~~~

**Comment by osrf-migration** [comment]
**Original comment by David Conner (Bitbucket: [dcconner](https://bitbucket.org/%7B4375dc4a-9371-4c51-973e-1f1d045590c5%7D/)).**

----------------------------------------
This code is generally intended to prevent integral wind up in the same direction as the joint command is saturated, and provide continuous control as the system moves out of saturation.  If the very large step change in the opposite direction causes saturation the other way, then I could see this "inverting" the integral term, and causing overshooting if the integral gain is relatively small.

The large step changes sounds more like a bang/bang style of control; do you need the integral term at all in those conditions?  This code is only active if the integral term is non-zero and can be further limited by the i_limits.

Removing the clamping will cause a different set of control issues for those using integral control.  That said, since I first made this suggestion, we have moved away from using integral control and will do so in limited situations.

A compromise would be to have a parameter that allows the user to turn clamping on or off for the system

**Comment by osrf-migration** [comment]
**Original comment by GoRobotGo (Bitbucket: [GoRobotGo](https://bitbucket.org/%7B2fd8dac8-286d-461a-b7dd-dba99deea025%7D/)).**

----------------------------------------
It was pretty easy to hit this even with a step of 0.1 radians on the shoulder / shx joint.  p= 500, i = 100, i_limit = 170, d=20.  Given that it was so easy to hit with a relatively small step, I don't think it should always be on.  Your suggestion to allow disabling it seems very reasonable.

**Comment by osrf-migration** [comment]
**Original comment by GoRobotGo (Bitbucket: [GoRobotGo](https://bitbucket.org/%7B2fd8dac8-286d-461a-b7dd-dba99deea025%7D/)).**

----------------------------------------
A comment on the severity - we are able to compile libAtlasPlugin.so to work around this.

**Comment by osrf-migration** [comment]
**Original comment by John Hsu (Bitbucket: [hsu](https://bitbucket.org/%7B0a186eae-abf0-4514-a951-23db5eccc286%7D/), GitHub: [hsu](https://github.com/hsu)).**

----------------------------------------
Hi @gorobotgo,

can you share your diff of AtlasPlugin?

thanks.

**Comment by osrf-migration** [comment]
**Original comment by GoRobotGo (Bitbucket: [GoRobotGo](https://bitbucket.org/%7B2fd8dac8-286d-461a-b7dd-dba99deea025%7D/)).**

----------------------------------------
I have the diff below, but all I did was comment out the integral tie-back section.
If you added a bool to AtlasCommand you could globally enable/disable the integral tie-back in AtlasPlugin.cpp  (below is the diff for version 2.5.2)
diff AtlasPlugin.cpp AtlasPlugin.cpp.orig 
1386d1385
< 	/*
1398d1396
< 	*/

**Comment by osrf-migration** [comment]
**Original comment by GoRobotGo (Bitbucket: [GoRobotGo](https://bitbucket.org/%7B2fd8dac8-286d-461a-b7dd-dba99deea025%7D/)).**

----------------------------------------
Here is how I would implement it.  I used the inverted logic so that tie-back would be enabled if the bool was 0 / false.  I think this would be most likely to leave it on for current users who didn't initialize the value.  The only change in AtlasPlugin.cpp was the addition of the first if statement.  The message would need a new bool.

```
#!c++

	if (!this->atlasCommand.disableIntegralTieback)
	{
          // integral tie-back during control saturation if using integral gain
          if (!math::equal(forceClamped,forceUnclamped) &&
              !math::equal((double)this->atlasState.ki_position[i],0.0) )
          {
            // lock integral term to provide continuous control as system moves
            // out of staturation
            this->errorTerms[i].k_i_q_i = math::clamp(
              this->errorTerms[i].k_i_q_i + (forceClamped - forceUnclamped),
            static_cast<double>(this->atlasState.i_effort_min[i]),
            static_cast<double>(this->atlasState.i_effort_max[i]));
          }
	}
```

**Comment by osrf-migration** [comment]
**Original comment by John Hsu (Bitbucket: [hsu](https://bitbucket.org/%7B0a186eae-abf0-4514-a951-23db5eccc286%7D/), GitHub: [hsu](https://github.com/hsu)).**

----------------------------------------
is it reasonable to make sure `k_i_q_i` has the same sign as `forceClamped`?

~~~
    // integral tie-back during control saturation if using integral gain
    if (!math::equal(forceClamped,forceUnclamped) &&
        !math::equal((double)this->atlasState.ki_position[i],0.0) )
    {
      // lock integral term to provide continuous control as system moves
      // out of staturation.  But make sure k_i_q_i has the same sign
      // as forceClamped, i.e. k_i_q_i does not clamp the wrong direction.
      if (forceClamped > 0.0)
        this->errorTerms[i].k_i_q_i = math::clamp(
          this->errorTerms[i].k_i_q_i + (forceClamped - forceUnclamped),
          0.0,
          static_cast<double>(this->atlasState.i_effort_max[i]));
      else
        this->errorTerms[i].k_i_q_i = math::clamp(
          this->errorTerms[i].k_i_q_i + (forceClamped - forceUnclamped),
          static_cast<double>(this->atlasState.i_effort_min[i]),
          0.0);
    }
~~~

**Comment by osrf-migration** [comment]
**Original comment by GoRobotGo (Bitbucket: [GoRobotGo](https://bitbucket.org/%7B2fd8dac8-286d-461a-b7dd-dba99deea025%7D/)).**

----------------------------------------
That solves about 2/3 of the problem for us.  It no longer causes an inversion of the k_i_q_i term and a wild swing in the wrong direction.  So, it is definitely an improvement.  It still zeroes out the k_i_q_i term which can cause overshoot.

In an example case - the arm joints angles are zero (arm straight out at side).  The steady state k_i_q_i term for the l_arm_shx joint is about 60.  If you make a step change to 0.1 (down) the tie-back will be activated unless kd_position is less than 1.7.  The proposed tie-back will zero k_i_q_i.  Without the integral term holding the arm up, the down motion will overshoot and then recover as k_i_q_i comes back up to about 60.

I don't know enough about the normal use cases for tie-back to know if this change would impact that.

Our preference would be to have the ability to disable tie-back, but the proposed change certainly improves the behavior for us.

**Comment by osrf-migration** [comment]
**Original comment by John Hsu (Bitbucket: [hsu](https://bitbucket.org/%7B0a186eae-abf0-4514-a951-23db5eccc286%7D/), GitHub: [hsu](https://github.com/hsu)).**

----------------------------------------
the overshoot you describe is a proper response of a standard integral term; in fact, with the tie-back, the overshoot due to integral term should be less than the result of a standard integral term implementation without tie-back.

**Comment by osrf-migration** [comment]
**Original comment by GoRobotGo (Bitbucket: [GoRobotGo](https://bitbucket.org/%7B2fd8dac8-286d-461a-b7dd-dba99deea025%7D/)).**

----------------------------------------
In the case I describe the overshoot is increased due to tie-back.  The arm's weight is being supported by the integral term.  When a command is sent to lower the arm the integral tie-back zeroes out the integral term.  Both gravity and the actuator are accelerating the arm downward and it overshoots an order of magnitude more than if tie-back is disabled.  If tie-back is disabled the integral term reduces slightly during the short movement, but still continues to support the arm and their is little overshoot in lowering the arm.

**Comment by osrf-migration** [comment]
**Original comment by GoRobotGo (Bitbucket: [GoRobotGo](https://bitbucket.org/%7B2fd8dac8-286d-461a-b7dd-dba99deea025%7D/)).**

----------------------------------------
I would like to put some specific numbers to this discussion.  The gains for the shx joint are set to p= 500, i = 100, i_limit = 170, d=20 and all arm joint angles are set to zero.  A downward step of 0.1 radians is commanded to the shx joint.  With the current tie-back the arm will shoot up by 0.15 radians before going to the commanded position (bad as it goes the opposite direction of the command!).  With the proposed tie-back algorithm the joint overshoots to 0.25 radians before recovering to the commanded position (bad as the overshoot is larger than the entire commanded change).  With tie-back disabled (the desired case) the joint overshoots to 0.15 radians before settling to the desired state.  Current tie-back causes a nasty joint reversal in the case.  The proposed tie-back causes a 3x worse overshoot than the desired capability to disable tie-back.

**Comment by osrf-migration** [comment]
**Original comment by John Hsu (Bitbucket: [hsu](https://bitbucket.org/%7B0a186eae-abf0-4514-a951-23db5eccc286%7D/), GitHub: [hsu](https://github.com/hsu)).**

----------------------------------------
couple of options here:

  1. remove integral tie-back ([pull request #272](https://osrf-migration.github.io/drcsim-gh-pages/#!/osrf/drcsim/pull-requests/272)).
  1. make sure integral tie-back does not flip sign of integral force term `k_i_q_i` ([pull request #273](https://osrf-migration.github.io/drcsim-gh-pages/#!/osrf/drcsim/pull-requests/273)).

I'll merge which pull request with more votes.

John

**Comment by osrf-migration** [comment]
**Original comment by GoRobotGo (Bitbucket: [GoRobotGo](https://bitbucket.org/%7B2fd8dac8-286d-461a-b7dd-dba99deea025%7D/)).**

----------------------------------------
I vote for choice number 1.  (big surprise :-)

**Comment by osrf-migration** [comment]
**Original comment by Nate Koenig (Bitbucket: [Nathan Koenig](https://bitbucket.org/%7Bc862cdd9-fcc8-4419-9fc7-e20db14b8fcb%7D/)).**

----------------------------------------

* changed state from "new" to "resolved"

Resolved in [pull request #272](https://osrf-migration.github.io/drcsim-gh-pages/#!/osrf/drcsim/pull-requests/272)
